### PR TITLE
Fix crash inside JSX Closing tag

### PR DESCRIPTION
**Pull request by yuit**
Fix issue #5953 by giving its associated opening tag. Initially, the PR tried to address preventing adding the global symbols that is not in jSX into the opening tag. However, doing so cause some inconsistent so the change is rolled back.

### CODE REVIEW

File: src/services/services.ts [hunk 1]
```diff
@@ -3113,7 +3114,9 @@ namespace ts {
             }
             else if (isRightOfOpenTag) {
                 let tagSymbols = typeChecker.getJsxIntrinsicTagNames();
-                if (tryGetGlobalSymbols()) {
+                // If the currect cursor is inside JSX opening tag, the only meaningful completions are those of JSX.IntrinsicElements or users defined React.Component
+                // If the services can't find those symbols, then show nothing instead of including all the global symbols in the completion list.
+                if (tryGetGlobalSymbols(/*includeAllGlobalSymbols*/false)) {
```
Comment (DanielRosenwasser) [inline src/services/services.ts]: Space after comment

File: src/services/services.ts [hunk 2]
```diff
@@ -3113,7 +3114,9 @@ namespace ts {
             }
             else if (isRightOfOpenTag) {
                 let tagSymbols = typeChecker.getJsxIntrinsicTagNames();
-                if (tryGetGlobalSymbols()) {
+                // If the currect cursor is inside JSX opening tag, the only meaningful completions are those of JSX.IntrinsicElements or users defined React.Component
```
Comment (DanielRosenwasser) [inline src/services/services.ts]: the user-defined

File: src/services/services.ts [hunk 3]
```diff
@@ -3113,7 +3114,9 @@ namespace ts {
             }
             else if (isRightOfOpenTag) {
                 let tagSymbols = typeChecker.getJsxIntrinsicTagNames();
-                if (tryGetGlobalSymbols()) {
+                // If the currect cursor is inside JSX opening tag, the only meaningful completions are those of JSX.IntrinsicElements or users defined React.Component
+                // If the services can't find those symbols, then show nothing instead of including all the global symbols in the completion list.
```
Comment (DanielRosenwasser) [inline src/services/services.ts]: If we can't find

File: src/services/services.ts (hunk 1 above)
Comment (DanielRosenwasser) [inline src/services/services.ts]: What does "all" mean? Something is getting filtered out and this parameter name doesn't convey it well

File: src/compiler/checker.ts [hunk 4]
```diff
@@ -14637,7 +14639,9 @@ namespace ts {
                     location = location.parent;
                 }
 
-                copySymbols(globals, meaning);
+                if (includeAllGlobalSymbols) {
+                    copySymbols(globals, meaning);
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: I don't understand why we would do this; presumably this means that a React component defined at the global scope wouldn't be put in the completion list.

Comment (yuit) [inline src/compiler/checker.ts]: If there is React component in global scope, we will have found it here https://github.com/Microsoft/TypeScript/pull/6006#discussion-diff-47050513R3116

File: src/services/services.ts (hunk 1 above)
Comment (yuit) [inline src/services/services.ts]: I agree on naming. not sure what is the best name.  What I try to convey is that the boolen control whether you will add all global symbols from all the files into the completions list or not. This doesn't make sense for JSX opening tag

File: src/services/services.ts (hunk 2 above)
Comment (sandersn) [inline src/services/services.ts]: currect -> current

**Comment by DanielRosenwasser** [comment]
After some though, I've realized that this sort of thing is potentially obstructive to the way that users might write code. For instance, let's say I have

```tsx
var x =
    <div>
        <span>
    </
```

Let's say I'm actually simply intent on closing my `div` element. I'm going to try to come back to my `span` element later, but you're giving no suggestions to me. Now that impedes the flow as I type, expecting to be completed by `div`.

I think you should consider giving completions for all containing open tags who don't have an **appropriate** corresponding closing tag. That means that for:

```ts
<div>
  <span>
    <a>
  </ span
</ div>
```

It might be the case that the `a` is considered closed, but its matching closing tag is `span`. That means `a` probably still needs to be closed, so you should show it in the completion list. That also means that `div` and `span` also lack a matching tag (`<span>` gets closed with `</ div>`, and `<div>` is unclosed).

We should discuss this a bit tomorrow.

**Comment by RyanCavanaugh** [comment]
> Let's say I'm actually simply intent on closing my div element

This seems like an odd state to be in.

@bowdenk7 - Maybe we can get some people writing JSX in our next usability study so we can see what order people actually go in. That could be very informative.

**Comment by yuit** [comment]
@DanielRosenwasser yes. I don't think I got what you trying to say in the last part of the comment
and I agree with @RyanCavanaugh it is odd to come back to close the tag later. I can see more that you want to come back and type between tags. 

Also to clarify, this behavior is only when you don't have react.d.ts.

**Comment by DanielRosenwasser** [comment]
We only show the closing tag element when you don't have a valid JSX namespace? That seems fine then. I'm mainly trying to avoid a scenario where you have someone with extremely weird typing habits (like myself) where they end up fighting the editor.

**Comment by yuit** [comment]
@DanielRosenwasser yep, this behavior is only trigger when we don't have valid JSX name.

**Comment by yuit** [comment]
@DanielRosenwasser @RyanCavanaugh @sandersn Any more comments?

**Comment by DanielRosenwasser** [comment]
We're exposing this in our API, so we should discuss this with @mhegazy in a meeting, but other than that :+1:

**Comment by mhegazy** [comment]
:+1:

**Pull request merged**